src/pixel/gain_lose_alpha.rs

```diff
@@ -0,0 +1,146 @@
+}
+/// A pixel which can lose its alpha component.
+pub trait LoseAlpha: HetPixel {
+    /// The pixel type after losing its alpha component.
+    type LoseAlpha: HetPixel;
+
+    /// Returns the pixel type after losing its alpha component.
+    fn lose_alpha(self) -> Self::LoseAlpha;
+
+    /// Returns the pixel type after losing its alpha component and the lost alpha component if any.
+    fn lose_alpha_verbose(self) -> (Self::LoseAlpha, Option<Self::AlphaComponent>);
+}
+
+macro_rules! lower_upper {
+    ($lower:ident, $upper:ident, {$($color_bit:tt),*}, $alpha_bit:tt) => {
+        impl<T> GainAlpha for $lower<T> where T: PixelComponent {
+            type GainAlpha = $upper<T>;
+
+            fn gain_alpha(self) -> Self::GainAlpha {
+                $upper {
+                    $($color_bit: self.$color_bit),*,
+                    $alpha_bit: <$lower<T> as HetPixel>::AlphaComponent::COMPONENT_MAX,
+                }
+            }
+            fn gain_alpha_with(self, alpha: Self::AlphaComponent) -> Self::GainAlpha {
+                $upper {
+                    $($color_bit: self.$color_bit),*,
+                    $alpha_bit: alpha,
+                }
+            }
+            fn gain_alpha_exact(self, alpha: Self::AlphaComponent) -> Self::GainAlpha {
+                $upper {
+                    $($color_bit: self.$color_bit),*,
+                    $alpha_bit: alpha,
+                }
+            }
+        }
+        impl<T> LoseAlpha for $upper<T> where T: PixelComponent {
+            type LoseAlpha = $lower<T>;
+
+            fn lose_alpha(self) -> Self::LoseAlpha {
+                $lower {
+                    $($color_bit: self.$color_bit),*
+                }
+            }
+            fn lose_alpha_verbose(self) -> (Self::LoseAlpha, Option<Self::AlphaComponent>) {
+                (
+                    $lower {
+                        $($color_bit: self.$color_bit),*
+                    },
+                    Some(self.$alpha_bit)
+                )
+            }
+        }
+    };
+}
+macro_rules! gain_already_alpha {
+    ($original:ident, $alpha_bit: tt) => {
+        impl<T> GainAlpha for $original<T>
+        where
+            T: PixelComponent,
+        {
+            type GainAlpha = $original<T>;
+
+            fn gain_alpha(self) -> Self::GainAlpha {
+                self
+            }
+            fn gain_alpha_with(self, _: Self::AlphaComponent) -> Self::GainAlpha {
+                self
+            }
+            fn gain_alpha_exact(mut self, alpha: Self::AlphaComponent) -> Self::GainAlpha {
+                self.$alpha_bit = alpha;
+                self
+            }
+        }
+    };
+}
+macro_rules! lose_already_no_alpha {
+    ($original:ident) => {
+        impl<T> LoseAlpha for $original<T>
+        where
+            T: PixelComponent,
+        {
+            type LoseAlpha = $original<T>;
+
+            fn lose_alpha(self) -> Self::LoseAlpha {
```

Comment:
I'm not sure if this is necessary. 

When removing alpha, you'd want some form of blending, because otherwise it uncovers invisible garbage. But doing that generically is a big task, because it'd need to know more about the color space. 

Maybe just fold these methods to pixel trait or HasAlpha?

Reply:
I'm not sure I understand, why would this have anything to do with blending? Doesn't blending require multiple pixels, like when combining two images?

We can't fold this into `HasAlpha` because then you wouldn't be able to call it on non-alpha pixels like `Rgb` which would limit generic functions applicability. `HetPixel` might work but it would require adding an extra associated type so I though it'd be nice to keep it separate as the complement to the also separate `GainAlpha` trait.

Reply:
I mean that if you just drop alpha channel and keep only RGB, then for alpha=0 you get garbage results.

For example it wouldn't be useful for converting a PNG to a JPEG, because it would uncover weird artifacts that weren't visible before. In practice to properly remove transparency you need to pick a background color (black, white, polka dots whatever). But the formula for blending with a bg differs for linear, sRGB, and premultiplied, and that's more detail than this crate wants to care about. So if it can't be done properly generically, let's leave it to application authors to do it themselves, and write dedicated code for RGBA or HasAlpha.

Reply:
Ah ok, I wasn't thinking about potential use-cases like converting pixel images with alpha to without alpha like `.png` to `.jpg` but now I think about it this would become a foot-gun so to speak. Good spot!

I'll remove the `LoseAlpha` trait, this is best implemented by users if truly required as you mentioned.